File: apps/mobile/src/grading/community/validation.ts
// Pure client-side validation for the community submission form.
//
// Mirrors the server-side rules in
// `apps/api-python/grading/flywheel/validation.py` so a submission that passes
// here also passes on the server (which re-validates — the client is never
// trusted). Keeping this pure (no React, no I/O) means the screen renders the
// error state from a value and the test-suite pins every rule without mounting.

import type {
  SubmitCommunitySubmissionRequest,
  CommunitySubgrades,
} from '@binderly/api-contracts';

import {
  SUBGRADE_KEYS,
  type CommunityGradeCompany,
  type CommunitySubmissionForm,
  type CommunitySubmissionImages,
  type SubgradeKey,
} from './types.js';

// Per-company cert-number shape, applied to the normalised digits-only cert.
// Heuristic ranges (not checksums) — they reject obvious typos, matching the
// Python `_CERT_PATTERNS`.
const CERT_PATTERNS: Record<CommunityGradeCompany, RegExp> = {
  PSA: /^\d{7,9}$/,
  BGS: /^\d{8,11}$/,
  CGC: /^\d{7,12}$/,
  SGC: /^\d{6,11}$/,
};

const COMPANY_TOKENS = ['PSA', 'BGS', 'BVG', 'CGC', 'SGC', 'BECKETT', 'CERT', '#'];

/**
 * Canonicalise a raw cert number: uppercase, strip a leading label token,
 * remove spaces / hyphens / `#`. Mirrors `normalize_cert_number` in Python.
 */
export function normalizeCertNumber(company: string, raw: string): string {
  let text = (raw ?? '').trim().toUpperCase();
  if (text.length === 0) return '';
  let changed = true;
  while (changed) {
    changed = false;
    for (const token of COMPANY_TOKENS) {
      if (text.startsWith(token)) {
        text = text.slice(token.length).replace(/^[\s:#-]+/, '');
        changed = true;
      }
    }
  }
  const companyUpper = (company ?? '').trim().toUpperCase();
  if (companyUpper.length > 0 && text.startsWith(companyUpper)) {
    text = text.slice(companyUpper.length).replace(/^[\s:#-]+/, '');
  }
  return text.replace(/[\s\-#]/g, '');
}

/** Parse a grade text field. Returns `null` for blank, `NaN` for non-numeric. */
function parseGrade(raw: string): number | null {
  const trimmed = (raw ?? '').trim();
  if (trimmed.length === 0) return null;
  return Number(trimmed);
}

/** A grade is valid iff it sits on the 0.5 grid within [1, 10]. */
export function isGradeOnGrid(value: number): boolean {
  if (!Number.isFinite(value)) return false;
  if (value < 1 || value > 10) return false;
  return Number.isInteger(value * 2);
}

export interface ValidationOutcome {
  readonly ok: boolean;
  /** Field-keyed error messages (e.g. `{ certNumber: '...' }`). */
  readonly errors: Readonly<Record<string, string>>;
  /** The wire-ready request — present only when `ok === true`. */
  readonly request?: SubmitCommunitySubmissionRequest;
}

/**
 * Validate the form + the backing photos and, on success, produce the
 * `SubmitCommunitySubmissionRequest` ready to POST.
 */
export function validateCommunityForm(
  form: CommunitySubmissionForm,
  images: CommunitySubmissionImages,
): ValidationOutcome {
  const errors: Record<string, string> = {};

  // ── cert number ──────────────────────────────────────────────────────────
  const normalized = normalizeCertNumber(form.gradeCompany, form.certNumber);
  if (normalized.length === 0) {
    errors.certNumber = 'Enter the cert number from the slab.';
  } else if (!CERT_PATTERNS[form.gradeCompany].test(normalized)) {
    errors.certNumber = `That doesn't look like a valid ${form.gradeCompany} cert number.`;
  }

  // ── overall grade ─────────────────────────────────────────────────────────
  const overall = parseGrade(form.overallGrade);
  if (overall !== null && !isGradeOnGrid(overall)) {
    errors.overallGrade = 'Grade must be between 1 and 10 in 0.5 steps.';
  }

  // ── sub-grades (optional) ──────────────────────────────────────────────────
  const subgrades: Record<string, number> = {};
  for (const key of SUBGRADE_KEYS) {
    const parsed = parseGrade(form.subgrades[key]);
    if (parsed === null) continue;
    if (!isGradeOnGrid(parsed)) {
      errors[`subgrade_${key}`] = `${capitalize(key)} sub-grade must be 1–10 in 0.5 steps.`;
    } else {
      subgrades[key] = parsed;
    }
  }

  // ── grade signal present? ───────────────────────────────────────────────────
  const hasSignal = form.blackLabel || overall !== null || Object.keys(subgrades).length > 0;
  if (!hasSignal) {
    errors.overallGrade = 'Enter the overall grade (or sub-grades / Black Label).';
  }

  // ── photos ────────────────────────────────────────────────────────────────
  if (!images.front) {
    errors.images = 'A front photo is required — capture the card first.';
  } else if (!images.back) {
    errors.images = 'A back photo is required — capture the card first.';
  }

  // ── consent ─────────────────────────────────────────────────────────────────
  if (!form.consent) {
    errors.consent = 'Please consent to your photos + grade being used as training data.';
  }

  if (Object.keys(errors).length > 0) {
    return { ok: false, errors };
  }

  const request: SubmitCommunitySubmissionRequest = {
    gradeCompany: form.gradeCompany,
    certNumber: form.certNumber.trim(),
    blackLabel: form.blackLabel,
    consent: true,
    images: buildImages(images),
    ...(overall !== null ? { overallGrade: overall } : {}),
    ...(Object.keys(subgrades).length > 0
      ? { subgrades: subgrades as CommunitySubgrades }
      : {}),
  };

  return { ok: true, errors: {}, request };
}

function buildImages(
  images: CommunitySubmissionImages,
): SubmitCommunitySubmissionRequest['images'] {
  // front + back are guaranteed present by the time we reach here.
  const out: {
    front: string;
    back: string;
    corners?: string[];
    surface?: string;
    slab?: string;
  } = {
    front: images.front as string,
    back: images.back as string,
  };
  if (images.corners && images.corners.length > 0) {
    out.corners = [...images.corners];
  }
  if (images.surface) out.surface = images.surface;
  if (images.slab) out.slab = images.slab;
  return out;
}

function capitalize(value: SubgradeKey): string {
  return value.charAt(0).toUpperCase() + value.slice(1);
}

File: apps/api-python/grading/flywheel/validation.py
"""Validation for community submissions.

Server-side mirror of the mobile client-side checks (apps/mobile/src/grading/
community/validation.ts). Both are pinned to the same per-company rules so a
submission that passes on-device also passes here — but the server is the
authority and re-validates everything (never trust the client).

What we validate:

1. **Company** — must be one of PSA / BGS / CGC / SGC.
2. **Cert-number format** — per-company heuristic on the *normalised* cert.
   These are shape checks (digit-count ranges), NOT authoritative checksum
   validation — PSA/Beckett/CGC do not publish a public checksum. The intent is
   to reject obvious typos (letters, empty, absurd lengths), not to guarantee
   the cert exists.
3. **Grade range** — overall + any sub-grades must be on the 0.5 grid in
   [1, 10], OR omitted (``None``) for a non-numeric outcome.
4. **Required photos** — at minimum a front + back reference (the centering +
   corners models need both). The slab photo is optional (some users won't
   share it; rules/07-grading.md surfaces it as optional).
5. **Consent** — the user must explicitly opt in to their data becoming
   training data.

``validate_submission`` returns a ``ValidationResult`` aggregating every error
rather than raising on the first, so the UI/caller can surface all problems at
once.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional

from .normalization import CANONICAL_GRADE_STEP, normalize_cert_number
from .types import CommunitySubmission, SUBGRADE_KEYS

SUPPORTED_COMPANIES: tuple[str, ...] = ("PSA", "BGS", "CGC", "SGC")

# Per-company cert-number shape (applied to the normalised, digits-only cert).
# Heuristic digit-count ranges drawn from observed real cert numbers; see the
# module docstring on why these are shape checks, not checksums.
_CERT_PATTERNS: dict[str, re.Pattern[str]] = {
    "PSA": re.compile(r"^\d{7,9}$"),
    "BGS": re.compile(r"^\d{8,11}$"),
    "CGC": re.compile(r"^\d{7,12}$"),
    "SGC": re.compile(r"^\d{6,11}$"),
}

_GRADE_MIN = 1.0
_GRADE_MAX = 10.0


class ValidationError(str):
    """A single human-readable validation failure code/message.

    Subclasses ``str`` so errors compare/serialise as plain strings while still
    being a distinct, greppable type at call sites.
    """


@dataclass(frozen=True)
class ValidationResult:
    """Aggregate validation outcome.

    ``ok`` is True iff ``errors`` is empty. ``errors`` are stable, prefixed
    codes (e.g. ``"cert_number: ..."``) so tests can assert on the field
    without coupling to exact prose.
    """

    ok: bool
    errors: list[str] = field(default_factory=list)

    @property
    def error_fields(self) -> set[str]:
        """The set of field prefixes that failed (text before the first ':')."""
        return {e.split(":", 1)[0] for e in self.errors}


def _grade_in_range(value: float) -> bool:
    if not (_GRADE_MIN <= value <= _GRADE_MAX):
        return False
    # On the 0.5 grid: 2*value must be an integer.
    doubled = value / CANONICAL_GRADE_STEP
    return abs(doubled - round(doubled)) < 1e-9


def validate_company(company: str) -> Optional[str]:
    if (company or "").strip().upper() not in SUPPORTED_COMPANIES:
        return f"grade_company: unsupported company {company!r} (expected one of {SUPPORTED_COMPANIES})"
    return None


def validate_cert_number(company: str, cert_number: str) -> Optional[str]:
    company_upper = (company or "").strip().upper()
    normalized = normalize_cert_number(company_upper, cert_number)
    if not normalized:
        return "cert_number: empty after normalisation"
    pattern = _CERT_PATTERNS.get(company_upper)
    if pattern is None:
        # Unknown company already reported by validate_company; nothing to add.
        return None
    if not pattern.fullmatch(normalized):
        return (
            f"cert_number: {cert_number!r} (normalised {normalized!r}) does not "
            f"match the expected {company_upper} format"
        )
    return None


def validate_grades(submission: CommunitySubmission) -> list[str]:
    errors: list[str] = []
    overall = submission.overall_grade
    # Black Label is implicitly a perfect 10; an explicit numeric overall is
    # still allowed but if present must be in range.
    if overall is not None and not _grade_in_range(overall):
        errors.append(
            f"overall_grade: {overall} is outside the [1, 10] 0.5-step grid"
        )
    # A non-Black-Label submission must carry *some* grade signal (an overall
    # or at least one sub-grade) — otherwise the row is unlabelled and useless.
    has_signal = (
        submission.black_label
        or overall is not None
        or bool(submission.subgrades)
    )
    if not has_signal:
        errors.append("overall_grade: no grade provided (overall or sub-grades required)")
    if submission.subgrades:
        for key, value in submission.subgrades.items():
            if key not in SUBGRADE_KEYS:
                errors.append(f"subgrades: unknown sub-grade {key!r}")
                continue
            if not _grade_in_range(value):
                errors.append(
                    f"subgrades: {key}={value} is outside the [1, 10] 0.5-step grid"
                )
    return errors


def validate_photos(submission: CommunitySubmission) -> list[str]:
    errors: list[str] = []
    images = submission.image_urls or {}
    front = images.get("front")
    back = images.get("back")
    if not isinstance(front, str) or not front:
        errors.append("image_urls: a front photo reference is required")
    if not isinstance(back, str) or not back:
        errors.append("image_urls: a back photo reference is required")
    corners = images.get("corners")
    if corners is not None and not isinstance(corners, list):
        errors.append("image_urls: corners must be a list of references when provided")
    return errors


def validate_consent(submission: CommunitySubmission) -> Optional[str]:
    if not submission.consent:
        return "consent: the user must consent to their data becoming training data"
    return None


def validate_submission(submission: CommunitySubmission) -> ValidationResult:
    """Run every check and aggregate the failures.

    Returns ``ValidationResult(ok=True, errors=[])`` for a fully valid
    submission. The order of checks is stable for deterministic test assertions.
    """
    errors: list[str] = []

    company_error = validate_company(submission.grade_company)
    if company_error:
        errors.append(company_error)

    # Only run the format check when the company is known (otherwise the
    # message is noise atop the company error).
    if company_error is None:
        cert_error = validate_cert_number(submission.grade_company, submission.cert_number)
        if cert_error:
            errors.append(cert_error)

    errors.extend(validate_grades(submission))
    errors.extend(validate_photos(submission))

    consent_error = validate_consent(submission)
    if consent_error:
        errors.append(consent_error)

    return ValidationResult(ok=not errors, errors=errors)
